### PR TITLE
crimson/os/seastore: promote read data to the faster device 

### PR DESCRIPTION
**Pull request by zhscn**
This PR implements the promotion as a part of the read cache based on the previous discussion. The basic idea is to rewrite the extent from LRU to the main device.

The next step might be:
1. Preventing cache pollution by replacing the current LRU algorithm with the ARC algorithm, as outlined in previous documentation.
2. Modifying the current GC/eviction policy to suit the promoted extent.

This PR relies on https://github.com/ceph/ceph/pull/47974

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/dev/crimson/crimson.rst [hunk 1]
```diff
@@ -182,6 +182,23 @@ using ``vstart.sh``,
 ``--memstore``
     use the alienized MemStore as the object store backend.
 
+``--seastore``
+    use the SeaStore as object store backend.
```
Comment (anthonyeleven) [inline doc/dev/crimson/crimson.rst:186]: Use SeaStore as the back end object store.

File: doc/dev/crimson/crimson.rst [hunk 2]
```diff
@@ -182,6 +182,23 @@ using ``vstart.sh``,
...
+
+``--seastore-devs``
+    specify the block device used by SeaStore.
```
Comment (anthonyeleven) [inline doc/dev/crimson/crimson.rst:189]: Specify capitalized.  Also s/use/Use/ on line 183.

File: doc/dev/crimson/crimson.rst [hunk 3]
```diff
@@ -182,6 +182,23 @@ using ``vstart.sh``,
...
+
+``--seastore-secondary-devs``
+    Optional.  SeaStore supports multiple devices.  Enable this feature by
+    passing the block device to this option.
+
+``--seastore-secondary-devs-type``
+    Optional.  Specify device type of secondary devices.  When the secondary
+    device is slower than main device passed to ``--seastore-devs``, the cold
+    data in faster device will be evicted to the slower devices over time.
+    Valid args include: HDD, SSD(default), ZNS and RANDOM_BLOCK_SSD.
```
Comment (anthonyeleven) [inline doc/dev/crimson/crimson.rst:199]: Uppercase CLI arguments are unusual but I don't think I can ask you to lowercase them in the code.

Suggest backticks around the literals, and a brief explantion of how `SSD` and `RANDOM_BLOCK_SSD` differ; on the surface the two seem to be identical.

Valid types include \`\`HDD\`\`, \`\`SSD\`\`(default), \`\`ZNS\`\`, and \`\`RANDOM
_BLOCK_SSD\`\``

File: src/common/options/crimson.yaml.in [hunk 4]
```diff
@@ -102,3 +107,18 @@ options:
...
   default: 5_G
+- name: seastore_multiple_tiers_stop_evict_ratio
+  type: float
+  level: advanced
+  desc: When the used ratio of main tier is less than this value, then stop evict cold data to the cold tier.
+  default: 0.5
+- name: seastore_multiple_tiers_default_evict_ratio
+  type: float
+  level: advanced
+  desc: start to evict cold data to the cold tier when the used ratio of main tier reach this ratio.
```
Comment (anthonyeleven) [inline src/common/options/crimson.yaml.in:118]: Begin evicting cold data to the cold tier when the used ratio of the main tier reaches this value.

File: src/common/options/crimson.yaml.in [hunk 5]
```diff
@@ -102,3 +107,18 @@ options:
...
+  default: 0.6
+- name: seastore_multiple_tiers_fast_evict_ratio
+  type: float
+  level: advanced
+  desc: start fast eviction when the used ratio of main tier reach this ratio.
```
Comment (anthonyeleven) [inline src/common/options/crimson.yaml.in:123]: Begin fast eviction when the used ratio of the main tier reaches this value.

File: src/crimson/os/seastore/cache.cc [hunk 6]
```diff
@@ -691,6 +730,16 @@ void Cache::register_metrics()
...
       ),
+      sm::make_counter(
+        "version_count_promotion",
+        stats.committed_promotion_version.num,
+        sm::description("total number of rewrite-promotion extents")
+      ),
+      sm::make_counter(
+        "version_sum_promotion",
+        stats.committed_promotion_version.version,
+        sm::description("sum of the version from rewrite-promotion extents")
```
Comment (anthonyeleven) [inline src/crimson/os/seastore/cache.cc:741]: Maybe `sum of the versions of rewrite-promotion extents` ?

File: src/vstart.sh [hunk 7]
```diff
@@ -255,7 +256,8 @@ options:
 	--no-restart: dont restart process when using ceph-run
 	--jaeger: use jaegertracing for tracing
 	--seastore-devs: comma-separated list of blockdevs to use for seastore
-	--seastore-secondary-des: comma-separated list of secondary blockdevs to use for seastore
+	--seastore-secondary-devs: comma-separated list of secondary blockdevs to use for seastore
+	--seastore-secondary-devs-type: device type of all secondary blockdevs, SSD or HDD
```
Comment (anthonyeleven) [inline src/vstart.sh:260]: Shouldn't this list match `DD, SSD(default), ZNS and RANDOM_BLOCK_SSD.` on line 199 of `crimson.rst`?

**Review by anthonyeleven — CHANGES_REQUESTED** [review]
I've requested a number of small docs-type changes.

**Comment by zhscn** [comment]
@anthonyeleven Thanks for your suggestion on the doc! The modifications of the document are from https://github.com/ceph/ceph/pull/47974. I'll merge the updates into that pull request.

File: src/vstart.sh (hunk 7 above)
Comment (zhscn) [reply src/vstart.sh:260]: Yeah, you are right, they should be consistent.

**Review by anthonyeleven — APPROVED** [review]

**Pull request closed without merging**